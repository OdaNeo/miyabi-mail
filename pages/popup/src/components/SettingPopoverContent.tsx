import { Button } from '@src/components/ui/button';
import { AlertTriangle, EyeOff, Eye, Trash2 } from 'lucide-react';
import { useRef, useState } from 'react';
import { Input } from '@src/components/ui/input';
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from './ui/select';
import { useStorage } from '@extension/shared';
import { apiKeyStorage, apiVersionStorage } from '@extension/storage';
import type React from 'react';
import { useI18n } from '@src/hooks/useI18n';
import { CopyButton } from './CopyButton';

export const SettingPopoverContent = ({ setIsOpen }: { setIsOpen: (isOpen: boolean) => void }) => {
  const [showKey, setShowKey] = useState(false);

  const apiVersion = useStorage(apiVersionStorage);
  const apiKey = useStorage(apiKeyStorage);

  const inputRef = useRef<HTMLInputElement>(null);

  const { API_KEY_SETTING, PLEASE_SET_API_KEY, SAVE } = useI18n();

  const handleClosePopover = () => {
    setIsOpen(false);
  };

  const handlePaste = (e: React.ClipboardEvent<HTMLInputElement>) => {
    const clipboardData = e.clipboardData;
    if (clipboardData) {
      const pastedData = clipboardData.getData('text');
      apiKeyStorage.set(pastedData);
    }
  };

  const handleInputClick = () => {
    if (inputRef.current) {
      inputRef.current.select();
    }
  };

  const handleCopyKey = () => {
    navigator.clipboard.writeText(apiKey);
  };

  const handleDelete = () => {
    apiKeyStorage.set('');
  };

  return (
    <div className="space-y-2">
      <h3 className="font-medium text-slate-800 dark:text-slate-200">{API_KEY_SETTING}</h3>
      {!apiKey && (
        <div className="text-red-500 flex items-center text-xs">
          <AlertTriangle className="h-3 w-3 mr-1" />
          {PLEASE_SET_API_KEY}
        </div>
      )}
      <div className="relative">
        <Input
          ref={inputRef}
          type={showKey ? 'text' : 'password'}
          placeholder="OpenAI API Key"
          value={apiKey}
          readOnly
          onPaste={handlePaste}
          onClick={handleInputClick}
          className="pr-20 bg-slate-100 dark:bg-slate-800 text-slate-800 dark:text-slate-200 placeholder-slate-500 dark:placeholder-slate-400 cursor-pointer"
        />
        <div className="absolute right-2 top-1/2 -translate-y-1/2 flex items-center gap-1">
          <Button
            variant="ghost"
            size="icon"
            data-testid="eye-icon"
            className="h-8 w-8 text-slate-500 hover:text-slate-900"
            onClick={() => setShowKey(!showKey)}
          >
            {showKey ? <EyeOff className="h-4 w-4" /> : <Eye className="h-4 w-4" />}
          </Button>
          {apiKey && <CopyButton handleCopyText={handleCopyKey} />}
        </div>
      </div>
      <Select value={apiVersion} onValueChange={apiVersionStorage.set}>
        <SelectTrigger className="w-full bg-slate-100 dark:bg-slate-800 text-slate-800 dark:text-slate-200">
          <SelectValue placeholder="Select API Version" />
        </SelectTrigger>
        <SelectContent>
          <SelectItem value="gpt-4o-mini">GPT-4o-mini</SelectItem>
          <SelectItem value="gpt-4o">GPT-4o</SelectItem>
          <SelectItem value="o1-mini">GPT-o1-mini</SelectItem>
          <SelectItem value="o1">GPT-o1</SelectItem>
        </SelectContent>
      </Select>
      <div className="flex gap-2">
        <Button
          disabled={!apiKey}
          data-testid="save-icon"
          className="flex-1 bg-emerald-500 text-white hover:bg-emerald-600 disabled:bg-slate-300 dark:disabled:bg-slate-700"
          onClick={handleClosePopover}
        >
          {SAVE}
        </Button>
        {apiKey && (
          <Button data-testid="delete-api-icon" variant="destructive" size="icon" onClick={handleDelete}>
            <Trash2 className="h-4 w-4" />
          </Button>
        )}
      </div>
    </div>
  );
};
