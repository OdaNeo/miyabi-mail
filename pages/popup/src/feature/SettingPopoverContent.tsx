import { Button, Input, Slider, Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from '@extension/ui';
import { AlertTriangle, EyeOff, Eye, Trash2 } from 'lucide-react';
import { useEffect, useRef, useState } from 'react';
import { useStorage } from '@extension/shared';
import { apiKeyStorage, apiVersionStorage, temperatureStorage } from '@extension/storage';
import type React from 'react';
import { useI18n } from '@/hooks/useI18n';
import { CopyButton } from '../components/CopyButton';
import { motion } from 'framer-motion';

export const SettingPopoverContent = ({ setIsOpen }: { setIsOpen: (isOpen: boolean) => void }) => {
  const [showKey, setShowKey] = useState(false);
  const [pasted, setPasted] = useState(false);

  const apiVersion = useStorage(apiVersionStorage);
  const apiKey = useStorage(apiKeyStorage);
  const temperatureFromStorage = useStorage(temperatureStorage);

  const [temperature, setTemperature] = useState(temperatureFromStorage);

  useEffect(() => {
    temperatureStorage.set(temperature);
  }, [temperature]);

  const inputRef = useRef<HTMLInputElement>(null);

  const { API_KEY_SETTING, PLEASE_SET_API_KEY, SAVE, PRECISE, CREATIVE, TEMPERATURE, BALANCE } = useI18n();

  const handleClosePopover = () => {
    setIsOpen(false);
  };

  const handlePaste = (e: React.ClipboardEvent<HTMLInputElement>) => {
    const clipboardData = e.clipboardData;
    if (clipboardData) {
      const pastedData = clipboardData.getData('text');
      apiKeyStorage.set(pastedData);

      setPasted(true);

      setTimeout(() => {
        setPasted(false);
      }, 2000);
    }
  };

  const handleInputClick = () => {
    if (inputRef.current) {
      inputRef.current.select();
    }
  };

  const handleCopyKey = () => {
    navigator.clipboard.writeText(apiKey);
  };

  const handleDelete = () => {
    apiKeyStorage.set('');
  };

  return (
    <div className="space-y-2">
      <h3 className="font-medium text-slate-800 dark:text-slate-200 text-sm">{API_KEY_SETTING}</h3>
      {!apiKey && (
        <div className="text-red-500 flex items-center">
          <AlertTriangle className="h-3 w-3 mr-1" />
          {PLEASE_SET_API_KEY}
        </div>
      )}
      <div className="relative">
        <Input
          ref={inputRef}
          type={showKey ? 'text' : 'password'}
          placeholder="OpenAI API Key"
          value={apiKey}
          readOnly
          onPaste={handlePaste}
          onClick={handleInputClick}
          className={`${apiKey ? 'pr-20' : 'pr-10'}  ${pasted && 'border-2 border-[#8b5cf6] focus-visible:ring-0'} 
            bg-slate-100 dark:bg-slate-800 text-slate-800 dark:text-slate-200 placeholder-slate-500 dark:placeholder-slate-400 cursor-pointer`}
        />
        <div className="absolute right-2 top-1/2 -translate-y-1/2 flex items-center gap-1">
          <Button
            variant="ghost"
            size="icon"
            data-testid="eye-icon"
            className="h-8 w-8 text-slate-500 hover:text-slate-900"
            onClick={() => setShowKey(!showKey)}
          >
            {showKey ? <EyeOff className="h-4 w-4" /> : <Eye className="h-4 w-4" />}
          </Button>
          {apiKey && <CopyButton handleCopyText={handleCopyKey} />}
        </div>
      </div>
      <Select value={apiVersion} onValueChange={apiVersionStorage.set}>
        <SelectTrigger className="w-full bg-slate-100 dark:bg-slate-800 text-slate-800 dark:text-slate-200">
          <SelectValue placeholder="Select API Version" />
        </SelectTrigger>
        <SelectContent>
          <SelectItem value="gpt-4o-mini">GPT-4o-mini</SelectItem>
          <SelectItem value="gpt-4o">GPT-4o</SelectItem>
          <SelectItem value="o1-mini">GPT-o1-mini</SelectItem>
          <SelectItem value="o1">GPT-o1</SelectItem>
        </SelectContent>
      </Select>
      <div className="space-y-4 !mt-3">
        <div className="flex items-center justify-between">
          <span className="text-sm font-medium">{TEMPERATURE}</span>
          <motion.span
            key={temperature}
            initial={{ opacity: 0, y: -10 }}
            animate={{ opacity: 1, y: 0 }}
            className="text-sm tabular-nums font-medium"
          >
            {temperature}
          </motion.span>
        </div>
        <div className="relative !mt-2">
          <Slider value={[temperature]} onValueChange={([value]) => setTemperature(value)} max={2} min={0} step={0.1} />
          <div className="flex justify-between mt-2 text-muted-foreground">
            <span>{PRECISE}</span>
            <span>{BALANCE}</span>
            <span>{CREATIVE}</span>
          </div>
        </div>
      </div>
      <div className="flex gap-2">
        <Button
          disabled={!apiKey}
          data-testid="save-icon"
          className="flex-1  text-white disabled:bg-slate-300 dark:disabled:bg-slate-700 
          bg-[#8b5cf6] hover:bg-[#7c3aed] dark:bg-[#a78bfa] dark:hover:bg-[#9061f9]"
          onClick={handleClosePopover}
        >
          {SAVE}
        </Button>
        {apiKey && (
          <Button
            data-testid="delete-api-icon"
            variant="outline"
            size="icon"
            className="border-[#f87171] dark:border-[#ef4444] text-[#ef4444] dark:text-[#f87171] hover:bg-[#fee2e2] dark:hover:bg-[#450a0a]/20 hover:text-[#dc2626] dark:hover:text-[#ef4444]"
            onClick={handleDelete}
          >
            <Trash2 className="h-4 w-4" />
          </Button>
        )}
      </div>
    </div>
  );
};
