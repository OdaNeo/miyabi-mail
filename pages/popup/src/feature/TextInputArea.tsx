import { Button, Textarea } from '@extension/ui';
import { motion, AnimatePresence } from 'framer-motion';
import { X } from 'lucide-react';
import { CopyButton } from '../components/CopyButton';
import { useI18n } from '@/hooks/useI18n';
import type { ChangeEvent, ClipboardEvent } from 'react';

export function TextInputArea({
  inputText,
  isLoading,
  handleInputTextChange,
  handleOnPaste,
  handleInputTextClear,
  handleCopyText,
}: {
  inputText: string;
  isLoading: boolean;
  handleInputTextChange: (e: ChangeEvent<HTMLTextAreaElement>) => void;
  handleOnPaste: (e: ClipboardEvent<HTMLTextAreaElement>) => void;
  handleInputTextClear: () => void;
  handleCopyText: () => void;
}) {
  const { INPUT_PLACEHOLDER } = useI18n();

  return (
    <div className="relative flex-grow group">
      <Textarea
        placeholder={INPUT_PLACEHOLDER}
        value={inputText}
        disabled={isLoading}
        data-testid="input-textarea"
        onChange={handleInputTextChange}
        onPaste={handleOnPaste}
        className="h-full resize-none text-sm scrollbar-custom 
      dark:bg-slate-800 dark:text-slate-100 dark:placeholder-slate-400 
      bg-white text-slate-900 placeholder-slate-500 pr-6"
      />
      {inputText && (
        <>
          <AnimatePresence>
            <motion.div
              initial={{ opacity: 0, scale: 0.8 }}
              animate={{ opacity: 1, scale: 1 }}
              exit={{ opacity: 0, scale: 0.8 }}
              className="absolute top-2 right-2"
            >
              <Button
                disabled={isLoading}
                data-testid="text-clear-icon"
                variant="ghost"
                size="icon"
                className="h-6 w-6"
                onClick={handleInputTextClear}
              >
                <X className="h-6 w-6" />
              </Button>
            </motion.div>
          </AnimatePresence>
          <div className="absolute bottom-2 right-2 opacity-0 group-hover:opacity-100 transition-opacity duration-200">
            <CopyButton handleCopyText={handleCopyText} />
          </div>
        </>
      )}
    </div>
  );
}
