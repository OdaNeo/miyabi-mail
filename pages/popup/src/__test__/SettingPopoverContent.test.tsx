import { render, fireEvent, screen, act } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { SettingPopoverContent } from '../feature/SettingPopoverContent';
import { useStorage } from '@extension/shared';
import { apiKeyStorage } from '@extension/storage';

class ResizeObserver {
  observe() {
    return null;
  }
  unobserve() {
    return null;
  }
  disconnect() {
    return null;
  }
}
global.ResizeObserver = ResizeObserver;

vi.mock('@extension/shared', () => ({
  useStorage: vi.fn(),
}));

vi.mock('@extension/storage', () => ({
  apiKeyStorage: { set: vi.fn() },
  apiVersionStorage: { set: vi.fn() },
  temperatureStorage: { set: vi.fn() },
}));

vi.mock('@/hooks/useI18n', () => ({
  useI18n: () => ({
    API_KEY_SETTING: 'API Key Setting',
    PLEASE_SET_API_KEY: 'Please set API key',
    SAVE: 'Save',
    PRECISE: 'Precise',
    CREATIVE: 'Creative',
    TEMPERATURE: 'Temperature',
    BALANCE: 'Balance',
  }),
}));

vi.mock(import('framer-motion'), async importOriginal => {
  const actual = await importOriginal();
  return {
    ...actual,
    div: ({ children, ...props }: any) => <div {...props}>{children}</div>,
    span: ({ children, ...props }: any) => <span {...props}>{children}</span>,
  };
});

const mockClipboard = {
  writeText: vi.fn().mockResolvedValue(undefined),
};

Object.assign(navigator, { clipboard: mockClipboard });

describe('SettingPopoverContent Component', () => {
  const mockSetIsOpen = vi.fn();
  const useStorageMock = vi.mocked(useStorage);
  const apiKeyStorageMock = vi.mocked(apiKeyStorage);

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('应该正确渲染没有 API key 的状态', () => {
    useStorageMock // 这个顺序与代码中的调用顺序有关
      .mockReturnValueOnce('gpt-4o') // apiVersion
      .mockReturnValueOnce('') // apiKey;
      .mockReturnValueOnce(0.7); // temperatureStorage;
    const { container } = render(<SettingPopoverContent setIsOpen={mockSetIsOpen} />);
    expect(container).toMatchSnapshot();
  });

  it('应该正确渲染有 API key 的状态', () => {
    useStorageMock
      .mockReturnValueOnce('gpt-4o') // apiVersion
      .mockReturnValueOnce('test-api-key') // apiKey;
      .mockReturnValueOnce(0.7); // temperatureStorage;
    const { container } = render(<SettingPopoverContent setIsOpen={mockSetIsOpen} />);
    expect(container).toMatchSnapshot();
  });

  it('should copy API key to clipboard when copy button is clicked', () => {
    useStorageMock
      .mockReturnValueOnce('gpt-4o') // apiVersion
      .mockReturnValueOnce('test-api-key') // apiKey;
      .mockReturnValueOnce(0.7); // temperatureStorage;
    render(<SettingPopoverContent setIsOpen={mockSetIsOpen} />);
    fireEvent.click(screen.getByTestId('copy-icon-group'));
    expect(mockClipboard.writeText).toHaveBeenCalledWith('test-api-key');
  });

  it('should delete API key when delete button is clicked', () => {
    useStorageMock
      .mockReturnValueOnce('gpt-4o') // apiVersion
      .mockReturnValueOnce('test-api-key') // apiKey;
      .mockReturnValueOnce(0.7); // temperatureStorage;
    render(<SettingPopoverContent setIsOpen={mockSetIsOpen} />);
    fireEvent.click(screen.getByTestId('delete-api-icon'));
    expect(apiKeyStorageMock.set).toHaveBeenCalledWith('');
  });

  it('应该在粘贴时正确设置 API key，并且UI要有变化', async () => {
    vi.useFakeTimers();
    useStorageMock
      .mockReturnValueOnce('gpt-4o') // apiVersion
      .mockReturnValueOnce('') // apiKey
      .mockReturnValueOnce(0.7); // temperatureStorage;
    const { getByPlaceholderText, container } = render(<SettingPopoverContent setIsOpen={mockSetIsOpen} />);
    const input = getByPlaceholderText('OpenAI API Key');
    fireEvent.paste(input, {
      clipboardData: {
        getData: () => 'test-pasted-api-key',
      },
    });
    expect(apiKeyStorageMock.set).toHaveBeenCalledTimes(1);
    expect(apiKeyStorageMock.set).toHaveBeenCalledWith('test-pasted-api-key');
    expect(input).toHaveClass('border-2 border-[#8b5cf6] focus-visible:ring-0');

    await act(async () => {
      vi.advanceTimersByTime(2000);
    });

    expect(container).not.toHaveClass('border-2 border-[#8b5cf6] focus-visible:ring-0');

    vi.useRealTimers();
  });

  it('should close popover when close button is clicked', () => {
    useStorageMock
      .mockReturnValueOnce('gpt-4o') // apiVersion
      .mockReturnValueOnce('test-api-key') // apiKey;
      .mockReturnValueOnce(0.7); // temperatureStorage;
    render(<SettingPopoverContent setIsOpen={mockSetIsOpen} />);
    fireEvent.click(screen.getByTestId('save-icon'));
    expect(mockSetIsOpen).toHaveBeenCalledWith(false);
  });

  it('should select input when click', () => {
    const mockSelect = vi.fn();
    useStorageMock
      .mockReturnValueOnce('gpt-4o') // apiVersion
      .mockReturnValueOnce('test-api-key') // apiKey;
      .mockReturnValueOnce(0.7); // temperatureStorage;
    const { getByPlaceholderText } = render(<SettingPopoverContent setIsOpen={mockSetIsOpen} />);
    const input = getByPlaceholderText('OpenAI API Key');
    (input as any).select = mockSelect;
    fireEvent.click(input);
    expect(mockSelect).toHaveBeenCalledTimes(1);
  });

  it('should toggle password visibility when eye icon is clicked', () => {
    useStorageMock
      .mockReturnValueOnce('gpt-4o') // apiVersion
      .mockReturnValueOnce('test-api-key') // apiKey
      .mockReturnValueOnce(0.7); // temperatureStorage;
    const { getByPlaceholderText, getByTestId } = render(<SettingPopoverContent setIsOpen={mockSetIsOpen} />);
    const input = getByPlaceholderText('OpenAI API Key');
    const eyeButton = getByTestId('eye-icon');
    expect(input).toHaveAttribute('type', 'password');
    fireEvent.click(eyeButton);
    expect(input).toHaveAttribute('type', 'text');
    fireEvent.click(eyeButton);
    expect(input).toHaveAttribute('type', 'password');
  });
});
