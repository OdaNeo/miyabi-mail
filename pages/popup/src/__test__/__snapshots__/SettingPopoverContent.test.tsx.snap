// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SettingPopoverContent Component > 应该正确渲染有 API key 的状态 1`] = `
<div>
  <div
    class="space-y-2"
  >
    <h3
      class="font-medium text-slate-800 dark:text-slate-200 text-sm"
    >
      API Key Setting
    </h3>
    <div
      class="relative"
    >
      <input
        class="flex h-9 w-full rounded-md border border-input px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm pr-20 false bg-slate-100 dark:bg-slate-800 text-slate-800 dark:text-slate-200 placeholder-slate-500 dark:placeholder-slate-400 cursor-pointer"
        placeholder="OpenAI API Key"
        readonly=""
        type="password"
        value="test-api-key"
      />
      <div
        class="absolute right-2 top-1/2 -translate-y-1/2 flex items-center gap-1"
      >
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent h-8 w-8 text-slate-500 hover:text-slate-900"
          data-testid="eye-icon"
        >
          <svg
            class="lucide lucide-eye h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2.062 12.348a1 1 0 0 1 0-.696 10.75 10.75 0 0 1 19.876 0 1 1 0 0 1 0 .696 10.75 10.75 0 0 1-19.876 0"
            />
            <circle
              cx="12"
              cy="12"
              r="3"
            />
          </svg>
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent h-6 w-6 text-slate-500 hover:text-slate-900 relative"
          data-testid="copy-icon-group"
        >
          <div
            class="absolute inset-0 flex items-center justify-center"
            style="opacity: 1; transform: none;"
          >
            <svg
              class="lucide lucide-copy h-4 w-4"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="14"
                rx="2"
                ry="2"
                width="14"
                x="8"
                y="8"
              />
              <path
                d="M4 16c-1.1 0-2-.9-2-2V4c0-1.1.9-2 2-2h10c1.1 0 2 .9 2 2"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
    <button
      aria-autocomplete="none"
      aria-controls="radix-:r6:"
      aria-expanded="false"
      class="flex h-9 items-center justify-between whitespace-nowrap rounded-md border border-input px-3 py-2 text-sm shadow-sm ring-offset-background data-[placeholder]:text-muted-foreground focus:outline-none focus:ring-1 focus:ring-ring disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 w-full bg-slate-100 dark:bg-slate-800 text-slate-800 dark:text-slate-200"
      data-state="closed"
      dir="ltr"
      role="combobox"
      type="button"
    >
      <span
        style="pointer-events: none;"
      >
        GPT-4o
      </span>
      <svg
        aria-hidden="true"
        class="lucide lucide-chevron-down h-4 w-4 opacity-50"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m6 9 6 6 6-6"
        />
      </svg>
    </button>
    <div
      class="space-y-4 !mt-3"
    >
      <div
        class="flex items-center justify-between"
      >
        <span
          class="text-sm font-medium"
        >
          Temperature
        </span>
        <span
          class="text-sm tabular-nums font-medium"
          style="opacity: 0; transform: translateY(-10px);"
        >
          0.7
        </span>
      </div>
      <div
        class="relative !mt-2"
      >
        <span
          aria-disabled="false"
          class="relative flex w-full touch-none select-none items-center"
          data-orientation="horizontal"
          dir="ltr"
          style="--radix-slider-thumb-transform: translateX(-50%);"
        >
          <span
            class="relative h-1.5 w-full grow overflow-hidden rounded-full bg-primary/20 bg-[#e5e7eb] dark:bg-[#2e3238]"
            data-orientation="horizontal"
          >
            <span
              class="absolute h-full bg-primary bg-[#8b5cf6] dark:bg-[#a78bfa]"
              data-orientation="horizontal"
              style="left: 0%; right: 65%;"
            />
          </span>
          <span
            style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(35% + 0px);"
          >
            <span
              aria-orientation="horizontal"
              aria-valuemax="2"
              aria-valuemin="0"
              aria-valuenow="0.7"
              class="block h-4 w-4 rounded-full border border-[#e5e7eb] dark:border-[#2e3238] border-primary/50 bg-background shadow transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 focus-visible:ring-[#8b5cf6]"
              data-orientation="horizontal"
              data-radix-collection-item=""
              role="slider"
              style=""
              tabindex="0"
            />
          </span>
        </span>
        <div
          class="flex justify-between mt-2 text-muted-foreground"
        >
          <span>
            Precise
          </span>
          <span>
            Balance
          </span>
          <span>
            Creative
          </span>
        </div>
      </div>
    </div>
    <div
      class="flex gap-2"
    >
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 shadow h-9 px-4 py-2 flex-1 text-white disabled:bg-slate-300 dark:disabled:bg-slate-700 bg-[#8b5cf6] hover:bg-[#7c3aed] dark:bg-[#a78bfa] dark:hover:bg-[#9061f9]"
        data-testid="save-icon"
      >
        Save
      </button>
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border bg-background shadow-sm h-9 w-9 border-[#f87171] dark:border-[#ef4444] text-[#ef4444] dark:text-[#f87171] hover:bg-[#fee2e2] dark:hover:bg-[#450a0a]/20 hover:text-[#dc2626] dark:hover:text-[#ef4444]"
        data-testid="delete-api-icon"
      >
        <svg
          class="lucide lucide-trash2 h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 6h18"
          />
          <path
            d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
          />
          <path
            d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
          />
          <line
            x1="10"
            x2="10"
            y1="11"
            y2="17"
          />
          <line
            x1="14"
            x2="14"
            y1="11"
            y2="17"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`SettingPopoverContent Component > 应该正确渲染没有 API key 的状态 1`] = `
<div>
  <div
    class="space-y-2"
  >
    <h3
      class="font-medium text-slate-800 dark:text-slate-200 text-sm"
    >
      API Key Setting
    </h3>
    <div
      class="text-red-500 flex items-center"
    >
      <svg
        class="lucide lucide-triangle-alert h-3 w-3 mr-1"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m21.73 18-8-14a2 2 0 0 0-3.48 0l-8 14A2 2 0 0 0 4 21h16a2 2 0 0 0 1.73-3"
        />
        <path
          d="M12 9v4"
        />
        <path
          d="M12 17h.01"
        />
      </svg>
      Please set API key
    </div>
    <div
      class="relative"
    >
      <input
        class="flex h-9 w-full rounded-md border border-input px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm pr-10 false bg-slate-100 dark:bg-slate-800 text-slate-800 dark:text-slate-200 placeholder-slate-500 dark:placeholder-slate-400 cursor-pointer"
        placeholder="OpenAI API Key"
        readonly=""
        type="password"
        value=""
      />
      <div
        class="absolute right-2 top-1/2 -translate-y-1/2 flex items-center gap-1"
      >
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent h-8 w-8 text-slate-500 hover:text-slate-900"
          data-testid="eye-icon"
        >
          <svg
            class="lucide lucide-eye h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2.062 12.348a1 1 0 0 1 0-.696 10.75 10.75 0 0 1 19.876 0 1 1 0 0 1 0 .696 10.75 10.75 0 0 1-19.876 0"
            />
            <circle
              cx="12"
              cy="12"
              r="3"
            />
          </svg>
        </button>
      </div>
    </div>
    <button
      aria-autocomplete="none"
      aria-controls="radix-:r0:"
      aria-expanded="false"
      class="flex h-9 items-center justify-between whitespace-nowrap rounded-md border border-input px-3 py-2 text-sm shadow-sm ring-offset-background data-[placeholder]:text-muted-foreground focus:outline-none focus:ring-1 focus:ring-ring disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 w-full bg-slate-100 dark:bg-slate-800 text-slate-800 dark:text-slate-200"
      data-state="closed"
      dir="ltr"
      role="combobox"
      type="button"
    >
      <span
        style="pointer-events: none;"
      >
        GPT-4o
      </span>
      <svg
        aria-hidden="true"
        class="lucide lucide-chevron-down h-4 w-4 opacity-50"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m6 9 6 6 6-6"
        />
      </svg>
    </button>
    <div
      class="space-y-4 !mt-3"
    >
      <div
        class="flex items-center justify-between"
      >
        <span
          class="text-sm font-medium"
        >
          Temperature
        </span>
        <span
          class="text-sm tabular-nums font-medium"
          style="opacity: 0; transform: translateY(-10px);"
        >
          0.7
        </span>
      </div>
      <div
        class="relative !mt-2"
      >
        <span
          aria-disabled="false"
          class="relative flex w-full touch-none select-none items-center"
          data-orientation="horizontal"
          dir="ltr"
          style="--radix-slider-thumb-transform: translateX(-50%);"
        >
          <span
            class="relative h-1.5 w-full grow overflow-hidden rounded-full bg-primary/20 bg-[#e5e7eb] dark:bg-[#2e3238]"
            data-orientation="horizontal"
          >
            <span
              class="absolute h-full bg-primary bg-[#8b5cf6] dark:bg-[#a78bfa]"
              data-orientation="horizontal"
              style="left: 0%; right: 65%;"
            />
          </span>
          <span
            style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(35% + 0px);"
          >
            <span
              aria-orientation="horizontal"
              aria-valuemax="2"
              aria-valuemin="0"
              aria-valuenow="0.7"
              class="block h-4 w-4 rounded-full border border-[#e5e7eb] dark:border-[#2e3238] border-primary/50 bg-background shadow transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 focus-visible:ring-[#8b5cf6]"
              data-orientation="horizontal"
              data-radix-collection-item=""
              role="slider"
              style=""
              tabindex="0"
            />
          </span>
        </span>
        <div
          class="flex justify-between mt-2 text-muted-foreground"
        >
          <span>
            Precise
          </span>
          <span>
            Balance
          </span>
          <span>
            Creative
          </span>
        </div>
      </div>
    </div>
    <div
      class="flex gap-2"
    >
      <button
        class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 shadow h-9 px-4 py-2 flex-1 text-white disabled:bg-slate-300 dark:disabled:bg-slate-700 bg-[#8b5cf6] hover:bg-[#7c3aed] dark:bg-[#a78bfa] dark:hover:bg-[#9061f9]"
        data-testid="save-icon"
        disabled=""
      >
        Save
      </button>
    </div>
  </div>
</div>
`;
